File: apps/negentropy/src/negentropy/knowledge/routes/catalog.py
"""Auto-extracted route module: Catalog management."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any
from uuid import UUID

from fastapi import APIRouter, HTTPException, Query, status
from pydantic import ValidationError  # noqa: F401

from negentropy.db.session import AsyncSessionLocal
from negentropy.knowledge._shared import (
    _build_document_response,
    _get_catalog_service,
    _resolve_user_display_names,
)
from negentropy.logging import get_logger

if TYPE_CHECKING:
    pass

# Lifecycle schema imports
from negentropy.knowledge.lifecycle_schemas import (  # noqa: F401
    AssignDocumentRequest,
    CatalogTreeResponse,
    CategorySuggestionResponse,
    DocumentProvenanceResponse,
    WikiEntryContentResponse,
    WikiNavTreeResponse,
    WikiPublishActionResponse,
)
from negentropy.knowledge.lifecycle_schemas import CatalogCreateRequest as _CatalogCreateReq
from negentropy.knowledge.lifecycle_schemas import CatalogNodeCreateRequest as _CatalogNodeCreateReq
from negentropy.knowledge.lifecycle_schemas import CatalogNodeResponse as _CatalogNodeResp
from negentropy.knowledge.lifecycle_schemas import CatalogNodeUpdateRequest as _CatalogNodeUpdateReq
from negentropy.knowledge.lifecycle_schemas import CatalogResponse as _CatalogResp
from negentropy.knowledge.lifecycle_schemas import CatalogUpdateRequest as _CatalogUpdateReq

logger = get_logger("negentropy.knowledge.api")
router = APIRouter()

# =============================================================================
# Phase 3: 文档目录编目 API
# =============================================================================


def _to_catalog_node_resp(row: dict, *, children_count: int = 0, document_count: int = 0) -> _CatalogNodeResp:
    """将 DAO 树查询行（dict）转换为 API 响应 Schema"""
    return _CatalogNodeResp(
        id=row["id"],
        catalog_id=row["catalog_id"],
        parent_id=row.get("parent_id"),
        name=row["name"],
        slug=row["slug"],
        node_type=row["node_type"],
        description=row.get("description"),
        sort_order=row["sort_order"],
        config=row.get("config") or {},
        document_id=row.get("document_id"),
        source_corpus_id=row.get("source_corpus_id"),
        depth=row.get("depth", 0),
        children_count=children_count,
        document_count=document_count,
    )


def _entry_orm_to_resp(
    entry: Any, *, depth: int = 0, children_count: int = 0, document_count: int = 0
) -> _CatalogNodeResp:
    """将 DocCatalogEntry ORM 对象转换为 API 响应 Schema"""
    from negentropy.knowledge.lifecycle.catalog_dao import _ENUM_TO_NODE_TYPE, _compute_slug

    return _CatalogNodeResp(
        id=entry.id,
        catalog_id=entry.catalog_id,
        parent_id=entry.parent_entry_id,
        name=entry.name,
        slug=_compute_slug(entry.name, entry.slug_override),
        node_type=_ENUM_TO_NODE_TYPE.get(entry.node_type, entry.node_type) if entry.node_type else "folder",
        description=entry.description,
        sort_order=entry.position or 0,
        config=entry.config or {},
        source_corpus_id=entry.source_corpus_id,
        depth=depth,
        children_count=children_count,
        document_count=document_count,
    )


def _build_update_kwargs(body: _CatalogNodeUpdateReq) -> dict[str, Any]:
    """以「请求中显式出现的字段」为 SSOT 构建 PATCH 更新 kwargs。

    关键不变量：显式传入的 ``parent_id=None``（提升为根节点）**必须保留**，
    严禁被 ``if v is not None`` 之类的 falsy 过滤吞掉——否则「移动到顶层」会
    静默丢失父指针更新，刷新后回退为子节点。``model_dump(exclude_unset=True)``
    已保证未传字段不出现于结果，因此无需再做 None 过滤。
    """
    return body.model_dump(exclude_unset=True)


def _catalog_orm_to_resp(catalog: Any) -> _CatalogResp:
    vis = catalog.visibility or "INTERNAL"
    return _CatalogResp(
        id=catalog.id,
        name=catalog.name,
        slug=catalog.slug,
        app_name=catalog.app_name,
        description=catalog.description,
        visibility=vis.lower() if isinstance(vis, str) else "INTERNAL",
        is_archived=catalog.is_archived or False,
        version=catalog.version or 1,
        owner_id=catalog.owner_id,
        config=catalog.config or {},
        created_at=catalog.created_at,
        updated_at=catalog.updated_at,
    )


# =============================================================================
# Phase 3 补全: /catalogs RESTful 路由（对标 BFF 代理约定）
# =============================================================================


def _build_tree_response(tree_data: list[dict]) -> CatalogTreeResponse:
    """复用：将 CTE 扁平列表转为 CatalogTreeResponse（含 children_count）"""
    id_to_children_count: dict[UUID, int] = {}
    for node in tree_data:
        pid = node.get("parent_id")
        if pid is not None:
            id_to_children_count[pid] = id_to_children_count.get(pid, 0) + 1
    items = [_to_catalog_node_resp(node, children_count=id_to_children_count.get(node["id"], 0)) for node in tree_data]
    max_depth = max((n.get("depth", 0) for n in tree_data), default=0) if tree_data else 0
    return CatalogTreeResponse(tree=items, total_nodes=len(items), max_depth=max_depth)


# --- Catalog CRUD ---


@router.get("/catalogs")
async def list_catalogs(
    app_name: str | None = Query(default=None),
    include_archived: bool = Query(default=False),
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=50, ge=1, le=200),
):
    """列出 Catalog（支持 app_name 过滤）"""
    catalog_svc = _get_catalog_service()
    async with AsyncSessionLocal() as db:
        catalogs, total = await catalog_svc.list_catalogs(
            db,
            app_name=app_name,
            include_archived=include_archived,
            offset=offset,
            limit=limit,
        )
    items = [_catalog_orm_to_resp(c) for c in catalogs]
    logger.info("api_list_catalogs", total=total)
    return {"items": items, "total": total, "offset": offset, "limit": limit}


@router.post("/catalogs")
async def create_catalog(body: _CatalogCreateReq) -> _CatalogResp:
    """创建 Catalog"""
    catalog_svc = _get_catalog_service()
    async with AsyncSessionLocal() as db:
        try:
            catalog = await catalog_svc.create_catalog(
                db,
                app_name=body.app_name,
                name=body.name,
                slug=body.slug,
                owner_id=body.owner_id,
                description=body.description,
                visibility=body.visibility.upper(),
                config=body.config if body.config else None,
            )
        except ValueError as exc:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
        await db.commit()
    logger.info("api_create_catalog", catalog_id=str(catalog.id))
    return _catalog_orm_to_resp(catalog)


@router.get("/catalogs/{catalog_id}")
async def get_catalog(catalog_id: UUID) -> _CatalogResp:
    """获取单个 Catalog 详情"""
    catalog_svc = _get_catalog_service()
    async with AsyncSessionLocal() as db:
        catalog = await catalog_svc.get_catalog(db, catalog_id)
    if catalog is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Catalog not found")
    logger.info("api_get_catalog", catalog_id=str(catalog_id))
    return _catalog_orm_to_resp(catalog)


@router.patch("/catalogs/{catalog_id}")
async def update_catalog(
    catalog_id: UUID,
    body: _CatalogUpdateReq,
) -> _CatalogResp:
    """更新 Catalog 属性"""
    catalog_svc = _get_catalog_service()
    update_kwargs = {k: v for k, v in body.model_dump(exclude_unset=True).items() if v is not None}
    if not update_kwargs:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="No fields to update")
    if "visibility" in update_kwargs:
        update_kwargs["visibility"] = update_kwargs["visibility"].upper()
    async with AsyncSessionLocal() as db:
        try:
            catalog = await catalog_svc.update_catalog(db, catalog_id, **update_kwargs)
        except ValueError as exc:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
        if catalog is None:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Catalog not found")
        await db.commit()
    logger.info("api_update_catalog", catalog_id=str(catalog_id))
    return _catalog_orm_to_resp(catalog)


@router.delete("/catalogs/{catalog_id}")
async def delete_catalog(catalog_id: UUID):
    """删除 Catalog（级联删除所有条目）"""
    catalog_svc = _get_catalog_service()
    async with AsyncSessionLocal() as db:
        deleted = await catalog_svc.delete_catalog(db, catalog_id)
        if not deleted:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Catalog not found")
        await db.commit()
    logger.info("api_delete_catalog", catalog_id=str(catalog_id))
    return {"detail": "Catalog deleted", "catalog_id": str(catalog_id)}


# --- Catalog Tree ---


@router.get("/catalogs/{catalog_id}/tree")
async def get_catalog_tree_v2(catalog_id: UUID) -> CatalogTreeResponse:
    """获取 Catalog 完整目录树"""
    catalog_svc = _get_catalog_service()
    async with AsyncSessionLocal() as db:
        tree_data = await catalog_svc.get_tree(db, catalog_id)
    resp = _build_tree_response(tree_data)
    logger.info("api_get_catalog_tree_v2", catalog_id=str(catalog_id), total_nodes=resp.total_nodes)
    return resp


# --- Catalog Entry CRUD ---


@router.get("/catalogs/{catalog_id}/entries")
async def list_catalog_entries(catalog_id: UUID) -> CatalogTreeResponse:
    """列出 Catalog 下所有条目"""
    catalog_svc = _get_catalog_service()
    async with AsyncSessionLocal() as db:
        tree_data = await catalog_svc.get_tree(db, catalog_id)
    resp = _build_tree_response(tree_data)
    logger.info("api_list_catalog_entries", catalog_id=str(catalog_id), total=len(resp.tree))
    return resp


@router.post("/catalogs/{catalog_id}/entries")
async def create_catalog_entry(
    catalog_id: UUID,
    body: _CatalogNodeCreateReq,
) -> _CatalogNodeResp:
    """创建目录条目（catalog_id 从路径获取）"""
    catalog_svc = _get_catalog_service()
    async with AsyncSessionLocal() as db:
        try:
            node = await catalog_svc.create_node(
                db,
                catalog_id=catalog_id,
                name=body.name,
                slug=body.slug,
                parent_id=body.parent_id,
                node_type=body.node_type,
                description=body.description,
                sort_order=body.sort_order,
                config=body.config if body.config else None,
            )
        except ValueError as exc:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
        await db.commit()
    logger.info("api_create_catalog_entry", node_id=str(node.id), catalog_id=str(catalog_id))
    return _entry_orm_to_resp(node)


@router.get("/catalogs/{catalog_id}/entries/{entry_id}")
async def get_catalog_entry(
    catalog_id: UUID,
    entry_id: UUID,
) -> _CatalogNodeResp:
    """获取单个目录条目详情"""
    catalog_svc = _get_catalog_service()
    async with AsyncSessionLocal() as db:
        node = await catalog_svc.get_node(db, entry_id)
    if node is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Catalog entry not found")
    logger.info("api_get_catalog_entry", entry_id=str(entry_id))
    return _entry_orm_to_resp(node)


@router.patch("/catalogs/{catalog_id}/entries/{entry_id}")
async def update_catalog_entry(
    catalog_id: UUID,
    entry_id: UUID,
    body: _CatalogNodeUpdateReq,
) -> _CatalogNodeResp:
    """更新目录条目属性"""
    catalog_svc = _get_catalog_service()
    async with AsyncSessionLocal() as db:
        update_kwargs = _build_update_kwargs(body)
        if not update_kwargs:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="No fields to update")
        try:
            node = await catalog_svc.update_node(db, entry_id, **update_kwargs)
        except ValueError as exc:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
        if node is None:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Catalog entry not found")
        await db.commit()
    logger.info("api_update_catalog_entry", entry_id=str(entry_id))
    return _entry_orm_to_resp(node)


@router.delete("/catalogs/{catalog_id}/entries/{entry_id}")
async def delete_catalog_entry(
    catalog_id: UUID,
    entry_id: UUID,
):
    """删除目录条目（级联删除子节点和文档关联）"""
    catalog_svc = _get_catalog_service()
    async with AsyncSessionLocal() as db:
        deleted = await catalog_svc.delete_node(db, entry_id)
        if not deleted:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Catalog entry not found")
        await db.commit()
    logger.info("api_delete_catalog_entry", entry_id=str(entry_id))
    return {"detail": "Catalog entry deleted", "entry_id": str(entry_id)}


# --- Catalog Documents ---


@router.get("/catalogs/{catalog_id}/documents")
async def get_catalog_documents(
    catalog_id: UUID,
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=200, ge=1, le=200),
):
    """获取 Catalog 作用域下可分配的候选文档列表

    语义：返回与 catalog 同 app_name 下 status='active' 的全部 KnowledgeDocument，
    供 UI 「添加文档到节点」对话框作为候选集。已归属文档由 UI 侧基于 existingDocIds
    灰出（见 AddDocumentsDialog.tsx / DocumentAssignmentSection.tsx）。

    跨 app 不可见：与 catalog_service.assign_document 的 app_name 同源断言对齐
    （ISSUE-011 Phase 3 不变量）。
    """
    from negentropy.models.perception import DocCatalog
    from negentropy.storage.service import DocumentStorageService

    async with AsyncSessionLocal() as db:
        catalog = await db.get(DocCatalog, catalog_id)
        if catalog is None:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail={"code": "CATALOG_NOT_FOUND", "message": "Catalog not found"},
            )
        catalog_app = catalog.app_name  # app_name 创建后不可变（perception.py DocCatalog 约束）

    storage_service = DocumentStorageService()
    docs, total = await storage_service.list_documents(
        corpus_id=None,
        app_name=catalog_app,
        limit=limit,
        offset=offset,
    )
    unique_user_ids = list({doc.created_by for doc in docs if doc.created_by})
    name_map = await _resolve_user_display_names(unique_user_ids)
    items = [_build_document_response(doc, name_map) for doc in docs]

    logger.info(
        "api_get_catalog_documents",
        catalog_id=str(catalog_id),
        total=total,
        app_name=catalog_app,
    )
    return {"items": items, "total": total, "offset": offset, "limit": limit}


# --- Entry Documents ---


@router.get("/catalogs/{catalog_id}/entries/{entry_id}/documents")
async def get_entry_documents(
    catalog_id: UUID,
    entry_id: UUID,
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=50, ge=1, le=200),
):
    """获取目录条目下已归属的文档列表（分页）"""
    catalog_svc = _get_catalog_service()
    async with AsyncSessionLocal() as db:
        documents, total = await catalog_svc.get_node_documents(db, entry_id, offset=offset, limit=limit)
    unique_user_ids = list({doc.created_by for doc in documents if doc.created_by})
    name_map = await _resolve_user_display_names(unique_user_ids)
    items = [_build_document_response(doc, name_map) for doc in documents]
    logger.info("api_get_entry_documents", entry_id=str(entry_id), total=total)
    return {"documents": items, "total": total, "offset": offset, "limit": limit}


@router.post("/catalogs/{catalog_id}/entries/{entry_id}/documents")
async def assign_documents_to_entry(
    catalog_id: UUID,
    entry_id: UUID,
    body: AssignDocumentRequest,
):
    """将一批文档归入目录条目（幂等操作）"""
    catalog_svc = _get_catalog_service()
    async with AsyncSessionLocal() as db:
        assigned_count = 0
        errors: list[str] = []
        for doc_id in body.document_ids:
            try:
                await catalog_svc.assign_document(db, entry_id, doc_id)
                assigned_count += 1
            except (ValueError, PermissionError) as exc:
                errors.append(f"{doc_id}: {exc}")
        await db.commit()
    logger.info("api_assign_documents_to_entry", entry_id=str(entry_id), assigned=assigned_count, errors=len(errors))
    result: dict[str, Any] = {"assigned_count": assigned_count, "total_requested": len(body.document_ids)}
    if errors:
        result["errors"] = errors
    return result


@router.delete("/catalogs/{catalog_id}/entries/{entry_id}/documents/{document_id}")
async def unassign_document_from_entry(
    catalog_id: UUID,
    entry_id: UUID,
    document_id: UUID,
):
    """从目录条目移除文档归属"""
    catalog_svc = _get_catalog_service()
    async with AsyncSessionLocal() as db:
        removed = await catalog_svc.unassign_document(db, entry_id, document_id)
        await db.commit()
    if not removed:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Document {document_id} not found in catalog entry {entry_id}",
        )
    logger.info("api_unassign_document_from_entry", entry_id=str(entry_id), document_id=str(document_id))
    return {"detail": "Document unassigned from catalog entry"}

File: apps/negentropy-ui/features/knowledge/utils/knowledge-api/catalog.ts
/**
 * 目录编册（Catalog）：目录树、节点与文档分配 API
 */
import type { KnowledgeDocument } from "./documents";

// ============================================================================
// Catalog Types (目录编册 — 对齐后端 catalog_dao.py)
// ============================================================================

/**
 * 目录节点类型（对齐后端 0010 收敛）：
 *   - `folder`：用户可见的目录容器（合并自历史 CATEGORY + COLLECTION）；
 *   - `document_ref`：系统内部软引用，仅由 `assign_document` 自动创建，UI 不暴露创建入口。
 *
 * 历史值 `category` / `collection` 仍可能从老 API 响应中返回，前端按 `folder` 兜底渲染。
 */
export type CatalogNodeType =
  | "folder"
  | "document_ref"
  | "category" // legacy
  | "collection"; // legacy

/** 全局 Catalog 元数据 — 对齐后端 DocCatalog ORM */
export interface DocCatalog {
  id: string;
  name: string;
  slug: string;
  app_name: string;
  description: string | null;
  visibility: "private" | "internal" | "public";
  is_archived: boolean;
  version: number;
  owner_id: string | null;
  created_at?: string;
  updated_at?: string;
}

export interface DocCatalogListResponse {
  items: DocCatalog[];
  total: number;
}

export interface DocCatalogDocumentsResponse {
  items: KnowledgeDocument[];
  total: number;
}

/** 目录节点 — 对齐后端 DocCatalogEntry + CTE 扩展字段 */
export interface CatalogNode {
  id: string;
  catalog_id: string;
  name: string;
  slug: string;
  parent_id: string | null;
  node_type: CatalogNodeType;
  description: string | null;
  sort_order: number;
  config: Record<string, unknown>;
  /** DOCUMENT_REF 叶子节点关联的文档 ID（FOLDER 等结构节点为 null） */
  document_id?: string | null;
  /** DOCUMENT_REF 关联文档所属 corpus（左栏重命名文档节点时据此直连 document API 改 display_name） */
  source_corpus_id?: string | null;
  /** CTE 计算字段：层级深度（根节点为 0） */
  depth?: number;
  /** CTE 计算字段：从根到当前节点的 ID 路径数组 */
  path?: string[];
  /** 前端派生：子节点数量 */
  children_count?: number;
  /** 前端派生：关联文档数量 */
  document_count?: number;
  created_at?: string;
  updated_at?: string;
}

export interface CreateCatalogNodeParams {
  catalog_id: string;
  name: string;
  slug: string;
  parent_id?: string | null;
  node_type?: CatalogNodeType;
  description?: string;
  sort_order?: number;
  config?: Record<string, unknown>;
}

export interface UpdateCatalogNodeParams {
  name?: string;
  slug?: string;
  parent_id?: string | null;
  node_type?: CatalogNodeType;
  description?: string;
  sort_order?: number;
  config?: Record<string, unknown>;
}

export interface CatalogTreeResponse {
  tree: CatalogNode[];
}

export interface CatalogNodesResponse {
  nodes: CatalogNode[];
  total: number;
}

export interface CatalogNodeDocumentsResponse {
  documents: KnowledgeDocument[];
  total: number;
}

// ============================================================================
// Catalog API Functions
// ============================================================================

/** 获取目录树（CTE 扁平化列表，含 depth/path） */
export async function fetchCatalogTree(catalogId: string): Promise<CatalogNode[]> {
  const res = await fetch(`/api/knowledge/catalogs/${catalogId}/tree`, {
    cache: "no-store",
  });
  if (!res.ok) throw new Error(`Failed to fetch catalog tree: ${res.statusText}`);
  const data = await res.json();
  return data.tree ?? data;
}

/** 获取目录节点列表（分页） */
export async function fetchCatalogNodes(params: {
  catalog_id: string;
  limit?: number;
  offset?: number;
}): Promise<CatalogNodesResponse> {
  const query = new URLSearchParams();
  if (params.limit != null) query.set("limit", String(params.limit));
  if (params.offset != null) query.set("offset", String(params.offset));
  const qs = query.toString();
  const res = await fetch(`/api/knowledge/catalogs/${params.catalog_id}/entries${qs ? `?${qs}` : ""}`, {
    cache: "no-store",
  });
  if (!res.ok) throw new Error(`Failed to fetch catalog nodes: ${res.statusText}`);
  return res.json();
}

/** 创建目录节点
 *
 * 后端契约（knowledge/api.py POST /catalogs/{catalog_id}/entries）：
 *   catalog_id 作为路径参数；body 中其他字段。
 */
export async function createCatalogNode(params: CreateCatalogNodeParams): Promise<CatalogNode> {
  const { catalog_id, ...body } = params;
  // 防御性校验：catalog_id 若为空字符串，模板字符串会降级出 `/api/knowledge/catalogs//entries`，
  // 经 Next.js URL 归一化后等效命中 `[catalogId]/route.ts`（catalogId="entries"），该路由无 POST 导致 405。
  // 在此前置显式报错，避免低可观测性的静默漂移。
  if (!catalog_id) {
    throw new Error("catalog_id is required to create a catalog node");
  }
  const res = await fetch(`/api/knowledge/catalogs/${catalog_id}/entries`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`Failed to create catalog node: ${res.statusText}`);
  return res.json();
}

/** 获取单个目录节点详情 */
export async function fetchCatalogNode(catalogId: string, nodeId: string): Promise<CatalogNode> {
  const res = await fetch(`/api/knowledge/catalogs/${catalogId}/entries/${nodeId}`, { cache: "no-store" });
  if (!res.ok) throw new Error(`Failed to fetch catalog node: ${res.statusText}`);
  return res.json();
}

/** 更新目录节点 */
export async function updateCatalogNode(
  catalogId: string,
  nodeId: string,
  params: UpdateCatalogNodeParams,
): Promise<CatalogNode> {
  const res = await fetch(`/api/knowledge/catalogs/${catalogId}/entries/${nodeId}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(params),
  });
  if (!res.ok) throw new Error(`Failed to update catalog node: ${res.statusText}`);
  return res.json();
}

/** 删除目录节点 */
export async function deleteCatalogNode(catalogId: string, nodeId: string): Promise<void> {
  const res = await fetch(`/api/knowledge/catalogs/${catalogId}/entries/${nodeId}`, { method: "DELETE" });
  if (!res.ok) throw new Error(`Failed to delete catalog node: ${res.statusText}`);
}

/** 获取目录节点下的文档列表（分页） */
export async function fetchCatalogNodeDocuments(
  catalogId: string,
  nodeId: string,
  options?: { limit?: number; offset?: number },
): Promise<CatalogNodeDocumentsResponse> {
  const query = new URLSearchParams();
  if (options?.limit != null) query.set("limit", String(options.limit));
  if (options?.offset != null) query.set("offset", String(options.offset));
  const qs = query.toString();
  const res = await fetch(`/api/knowledge/catalogs/${catalogId}/entries/${nodeId}/documents${qs ? `?${qs}` : ""}`, {
    cache: "no-store",
  });
  if (!res.ok) throw new Error(`Failed to fetch node documents: ${res.statusText}`);
  return res.json();
}

/** 将文档分配到目录节点（通过批量端点） */
export async function assignDocumentToNode(
  catalogId: string,
  nodeId: string,
  docId: string,
): Promise<void> {
  const res = await fetch(`/api/knowledge/catalogs/${catalogId}/entries/${nodeId}/documents`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ document_ids: [docId] }),
  });
  if (!res.ok) throw new Error(`Failed to assign document: ${res.statusText}`);
}

/** 从目录节点移除文档 */
export async function unassignDocumentFromNode(
  catalogId: string,
  nodeId: string,
  docId: string,
): Promise<void> {
  const res = await fetch(`/api/knowledge/catalogs/${catalogId}/entries/${nodeId}/documents/${docId}`, {
    method: "DELETE",
  });
  if (!res.ok) throw new Error(`Failed to unassign document: ${res.statusText}`);
}

/** 列出全局 Catalog（按 app_name 过滤） */
export async function fetchCatalogs(params?: {
  appName?: string;
  limit?: number;
  offset?: number;
}): Promise<DocCatalogListResponse> {
  const query = new URLSearchParams();
  if (params?.appName) query.set("app_name", params.appName);
  if (params?.limit != null) query.set("limit", String(params.limit));
  if (params?.offset != null) query.set("offset", String(params.offset));
  const qs = query.toString();
  const res = await fetch(`/api/knowledge/catalogs${qs ? `?${qs}` : ""}`, {
    cache: "no-store",
  });
  if (!res.ok) throw new Error(`Failed to fetch catalogs: ${res.statusText}`);
  return res.json();
}

/** 创建全局 Catalog */
export async function createCatalog(params: {
  app_name: string;
  name: string;
  slug: string;
  visibility?: string;
}): Promise<DocCatalog> {
  const res = await fetch("/api/knowledge/catalogs", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(params),
  });
  if (!res.ok) throw new Error(`Failed to create catalog: ${res.statusText}`);
  return res.json();
}

/** 获取 Catalog 下可用文档（跨 corpus，用于 AddDocumentsDialog） */
export async function fetchCatalogDocuments(
  catalogId: string,
  params?: { limit?: number; offset?: number; markdownStatus?: string },
): Promise<DocCatalogDocumentsResponse> {
  const query = new URLSearchParams();
  if (params?.limit != null) query.set("limit", String(params.limit));
  if (params?.offset != null) query.set("offset", String(params.offset));
  if (params?.markdownStatus) query.set("markdown_status", params.markdownStatus);
  const qs = query.toString();
  const res = await fetch(
    `/api/knowledge/catalogs/${catalogId}/documents${qs ? `?${qs}` : ""}`,
    { cache: "no-store" },
  );
  if (!res.ok) throw new Error(`Failed to fetch catalog documents: ${res.statusText}`);
  return res.json();
}

/**
 * 拉取 Catalog 下「全部」候选文档——以 200/页循环递增 offset 直至累计达到 total，
 * 突破后端单次 `le=200` 上限，保证「添加文档节点」候选集不被截断。
 *
 * @param markdownStatus 透传 `markdown_status` 过滤（UI 默认 "completed"：仅已转换为
 *   Markdown 的文档才可渲染为 Wiki 节点）；省略则不过滤。
 * 上限 100 页（20000 条）兜底，防异常 total 导致死循环。
 */
export async function fetchAllCatalogDocuments(
  catalogId: string,
  opts?: { markdownStatus?: string },
): Promise<KnowledgeDocument[]> {
  const PAGE = 200;
  const MAX_PAGES = 100;
  const acc: KnowledgeDocument[] = [];
  let offset = 0;
  for (let i = 0; i < MAX_PAGES; i++) {
    const res = await fetchCatalogDocuments(catalogId, {
      limit: PAGE,
      offset,
      markdownStatus: opts?.markdownStatus,
    });
    const items = res.items ?? [];
    acc.push(...items);
    const total = res.total ?? acc.length;
    if (acc.length >= total || items.length < PAGE) break;
    offset += PAGE;
  }
  return acc;
}
